taichi/system/timeline.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include <vector>
+#include <mutex>
+
+#include "taichi/common/core.h"
+#include "taichi/system/timer.h"
+
+TI_NAMESPACE_BEGIN
+
+struct TimelineEvent {
+  std::string name;
+  bool begin;
+  float64 time;
+  std::string tid;
+
+  std::string to_json();
+};
+
+class Timeline {
+ public:
+  Timeline();
+
+  ~Timeline();
+
+  static Timeline &get_this_thread_instance();
+
+  void set_name(const std::string &tid) {
+    tid_ = tid;
+  }
+
+  std::string get_name() {
+    return tid_;
+  }
+
+  void clear();
+
+  void insert_event(const TimelineEvent &e);
+
+  std::vector<TimelineEvent> fetch_events();
+
+  class Guard {
+   public:
+    Guard(const std::string &name);
+
+    ~Guard();
+
+   private:
+    std::string name_;
+  };
+
+ private:
+  std::string tid_;
+  std::mutex mut_;
```

Comment:
From what I understand, each `Timeline` is used in its own thread. So this is probably unnecessary?

Reply:
Originally I aimed for lock-free via the current thread-caching design, yet later I realized that in order to implement `clear` and `save`, the host thread needs to talk to each thread-local timeline, so we still need a lock :-( 

Let me know if you have better ideas on this :-)

According to my tests, the overhead of a mutex is not too much, as long as lock contention does not happen frequently.